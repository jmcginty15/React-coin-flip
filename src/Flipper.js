import { useState } from 'react';
import Coin from './Coin';

const Flipper = () => {
    const [flips, setFlips] = useState(0);
    const [heads, setHeads] = useState(0);
    const [tails, setTails] = useState(0);
    const [lastFlip, setLastFlip] = useState(null);

    return (
        <div className="Flipper">
            <h3>Let's flip a coin!</h3>
            <div>{lastFlip ? <Coin result={lastFlip} /> : null}</div>
            <button onClick={flipCoin}>FLIP MEEEE</button>
            <p>Out of {flips} flips, there have been {heads} heads and {tails} tails.</p>
        </div>
    )

    function flipCoin() {
        setFlips(flips + 1);
        const flipResult = Math.random() < 0.5 ? 'heads' : 'tails';

        if (flipResult === 'heads') {
            setLastFlip('heads');
            setHeads(heads + 1);
        } else if (flipResult === 'tails') {
            setLastFlip('tails');
            setTails(tails + 1);
        }
    }
}

export default Flipper;