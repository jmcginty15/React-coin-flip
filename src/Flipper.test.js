import React from "react";
import { render, fireEvent, queryByAltText } from "@testing-library/react";
import Flipper from "./Flipper";

beforeEach(function () {
    jest.spyOn(Math, "random")
        .mockReturnValueOnce(0.25)
        .mockReturnValueOnce(0.75);
});

// smoke test
it("renders without crashing", function () {
    render(<Flipper />);
});

// snapshot test
it("matches snapshot", function () {
    const { asFragment } = render(<Flipper />);
    expect(asFragment()).toMatchSnapshot();
});

it("does not show coin image on initial page load", function () {
    const { queryByAltText } = render(<Flipper />);
    expect(queryByAltText('heads')).not.toBeInTheDocument();
    expect(queryByAltText('tails')).not.toBeInTheDocument();
});

it("renders correct image for heads or tails result", function () {
    const { queryByAltText, queryByTestId } = render(<Flipper />);
    const flipButton = queryByTestId('flip-button');
    fireEvent.click(flipButton);
    expect(queryByAltText('heads')).toBeInTheDocument();
    expect(queryByAltText('tails')).not.toBeInTheDocument();
    fireEvent.click(flipButton);
    expect(queryByAltText('heads')).not.toBeInTheDocument();
    expect(queryByAltText('tails')).toBeInTheDocument();
});

it("updates flip count and result counts correctly", function () {
    const { queryByText, queryByTestId } = render(<Flipper />);
    const flipButton = queryByTestId('flip-button');
    fireEvent.click(flipButton);
    expect(queryByText('Out of 1 flips, there have been 1 heads and 0 tails.')).toBeInTheDocument();
    fireEvent.click(flipButton);
    expect(queryByText('Out of 2 flips, there have been 1 heads and 1 tails.')).toBeInTheDocument();
});

afterEach(function () {
    Math.random.mockRestore();
});
