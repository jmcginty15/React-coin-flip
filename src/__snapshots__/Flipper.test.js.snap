// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`matches snapshot 1`] = `
<DocumentFragment>
  <div
    class="Flipper"
  >
    <h3>
      Let's flip a coin!
    </h3>
    <div />
    <button
      data-testid="flip-button"
    >
      FLIP MEEEE
    </button>
    <p>
      Out of 0 flips, there have been 0 heads and 0 tails.
    </p>
  </div>
</DocumentFragment>
`;
